keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/README.md

```diff
@@ -0,0 +1,1246 @@
+
+PR: https://github.com/jpbetz/kubernetes/pull/61 \
+Github gist of changes: https://gist.github.com/aaron-prindle/b106b24f74770218b3e84005ddeb1bca
+
+```go
+        for _, tc := range successCases { // <--- errorCases done similarly
+                for _, gateVal := range []bool{false, true} {
+                        gate := features.DeclarativeValidation
+                        featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, gate, gateVal)
+
+                        errs := ValidateReplicationController(&tc, PodValidationOptions{})
+                        if utilfeature.DefaultFeatureGate.Enabled(gate) {
+                                // If declarative validation is enabled, it's the union of
+                                // managed and declarative validation that we are testing.
+                                versioned := v1.ReplicationController{}
+                                if err := v1util.Convert_core_ReplicationController_To_v1_ReplicationController(&tc, &versioned, nil); err != nil {
+                                        t.Fatalf("failed to convert to v1: %v", err)
+                                }
+                                ctx := request.WithRequestInfo(context.Background(), &request.RequestInfo{
+                                        APIGroup:   "",
+                                        APIVersion: "v1",
+                                })
+                                errs = append(errs, rest.ValidateDeclaratively(ctx, nil, legacyscheme.Scheme, &versioned)...)
+                        }
+                        if len(errs) != 0 {
+                                t.Errorf("expected success: %v", errs)
+                        }
+                }
+
+                verifyVersionedValidationEquivalence(t, &tc, nil)
+
+```
+
+###### Fuzz testing
+
+The current `validation_test.go` tests may have gaps such that using the current `validaton_test.go` tests for 1:1 functional equivalence is not always guaranteed.  For example, in the current Declarative Validation prototype migration experiment attempting to migrate ReplicationSpec ([here](https://github.com/jpbetz/kubernetes/pull/61)) it was discovered that the current tests did not validate all the validation logic of all fields properly.  In order to enhance the equivalency checks being done there will also be new fuzz tests added to `validation_test.go` with similar equivalency validations to those mentioned above.  This way we enhance the coverage of our equivalency testing and be more confident in our assumption that the hand-written vs declarative validation is logically identical.  An example of what this might look like can be seen in the snippet below:
```

Comment:
will the fuzzer make use of validation tags to fuzz both valid and invalid fields?  The value space of invalid values is likely larger than valid values.

Reply:
What we are doing now is fuzzing inputs and verifying the same errors come out in all versions, modulo field name (which I know is suspect and is on the backlog)